src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -553,35 +553,46 @@ class Wavefunction
     /// \pre: Each qubit, listed in `q`, must be unentangled and in state |0>.
     /// Place qubits, listed in `q` into superposition of basis vectors with provided `amplitudes`, where the order of
     /// qubits in array `q` defines the order of the basis vectors (little endian).
-    /// NB: at the moment supported only total state injection on all currently allocated qubits
     void inject_state(const std::vector<logical_qubit_id>& qubits, const std::vector<ComplexType>& amplitudes)
     {
         assert((static_cast<size_t>(1) << qubits.size()) == amplitudes.size());
 
-        if (num_qubits_ != qubits.size()) throw std::runtime_error("Subsystem state injection not supported (yet)");
-
         flush();
 
         // Check prerequisites.
         for (logical_qubit_id q : qubits)
         {
-            if (!kernels::isclassical(wfn_, get_qubit_position(q))
-                || kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
+            if (!kernels::isclassical(wfn_, get_qubit_position(q)) ||
+                kernels::getvalue(wfn_, get_qubit_position(q)) != 0)
             {
                 throw std::runtime_error("Cannot prepare state of entangled qubits or if they are not in state |0>");
             }
         }
 
-        // Reorder the qubits to the positions that match the wave function provided by the user.
-        for (unsigned i = 0; i < qubits.size(); i++)
+        if (qubits.size() == num_qubits_)
         {
-            qubitmap_[qubits[i]] = i;
+            // For full state injection we can simply copy the user's wave function into our store and reorder the
```

Comment:
This optimization should also hold if the subregister not being injected is in a single computational basis state. That could even possibly be extended further to use sparsity in the computational basis, perhaps?

Reply:
When injecting a partial state we don't know what the complementary qubits are doing and to find out that the current wave function is sparse would require us to traverse it. And even if the current wave function is just a single term we'd still have to compute the registers before assigning the new amplitudes. Basically, I don't believe there is an optimization opportunity here but please do reactivate if I've missed your point.